src/com/ms/silverking/cloud/meta/SuspectSet.java

```diff
@@ -0,0 +1,71 @@
+package com.ms.silverking.cloud.meta;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+import com.ms.silverking.collection.CollectionUtil;
+import com.ms.silverking.net.IPAndPort;
+
+public class SuspectSet extends ServerSetExtension {
+  private SuspectSet(ServerSet serverSet, long mzxid) {
+    super(serverSet);
+    this.mzxid = mzxid;
+  }
+
+  public SuspectSet(ServerSet serverSet) {
+    this(serverSet, INVALID_ZXID);
+  }
+
+  private SuspectSet(long version) {
+    this(new ServerSet(new HashSet<>(), version));
+  }
+
+  public SuspectSet(Set<String> excludedEntities, long version, long mzxid) {
+    this(new ServerSet(excludedEntities, version), mzxid);
+  }
+
+  public static SuspectSet emptySuspectSet(long version) {
+    return new SuspectSet(version);
+  }
+
+  @Override
+  public SuspectSet addByIPAndPort(Set<IPAndPort> newExcludedEntities) {
+    return (SuspectSet) super.addByIPAndPort(newExcludedEntities);
+  }
+
+  @Override
+  public SuspectSet add(Set<String> newExcludedEntities) {
+    return new SuspectSet(serverSet.add(newExcludedEntities));
+  }
+
+  @Override
+  public SuspectSet removeByIPAndPort(Set<IPAndPort> newExcludedEntities) {
+    return (SuspectSet) super.removeByIPAndPort(newExcludedEntities);
+  }
+
+  @Override
+  public SuspectSet remove(Set<String> newExcludedEntities) {
+    return new SuspectSet(serverSet.remove(newExcludedEntities));
+  }
+
+  public static SuspectSet parse(String def) {
+    return new SuspectSet(
+        new ServerSet(CollectionUtil.parseSet(def, singleLineDelimiter), VersionedDefinition.NO_VERSION));
+  }
+
+  public static SuspectSet parse(File file) throws IOException {
+    return new SuspectSet(ServerSet.parse(new FileInputStream(file), VersionedDefinition.NO_VERSION));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when ServerSet.parse throws an exception. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).